src/ucp/core/ucp_ep.c

```diff
@@ -910,12 +926,7 @@ static void ucp_ep_set_lanes_failed(ucp_ep_h ep, uct_ep_h *uct_eps)
     uct_ep_h uct_ep;
 
     ucp_ep_check_lanes(ep);
-
-    if (!(ep->flags & (UCP_EP_FLAG_FAILED | UCP_EP_FLAG_INTERNAL))) {
-        ucp_ep_release_id(ep);
-    }
-
-    ucp_ep_update_flags(ep, UCP_EP_FLAG_FAILED, 0);
+    ucp_ep_set_failed(ep);
```

Comment:
can we set FAILED flag in ucp_ep_close_nbx() instead of in ucp_ep_set_lanes_failed()?
and here - only assert that ep->flags&FAILED

Reply:
add comment that we release if here to drop incoming protocols

Reply:
move this to close_nb, and here only assert that FAILED

Reply:
/* Release ID here to prevent protocols from sending reply */

Reply:
> /* Release ID here to prevent protocols from sending reply */

done in `ucp_worker_set_ep_failed()`

Reply:
> move this to close_nb, and here only assert that FAILED

unfortunately, we can't do it only there - it should be done also for UCP EPs released in UCP Worker destroy.